scripts/gen_attribute_completion_and_shape_inference_hpp.py

```diff
@@ -0,0 +1,322 @@
+import os
+
+
+def make_completion_code(op_name,
+                         attribute_list=[],
+                         shape_inference_code='''
+assert(node.input_name_list.size() == 1);
+assert(node.output_name_list.size() == 1);
+add_variable_to_table(output(0), dtype_of(input(0)), dims_of(input(0)));
+''',
+                         preprocess="",
+                         postprocess=""):
+    # attribute completion and definition
+    attribute_completion_code_list = []
+    attribute_definition_list = []
+    for attribute in attribute_list:
+        attr_name, attr_type, default_value = attribute
+        inner_code = ''
+        if default_value is None:
+            inner_code = '''
+assert(!"attribute not found: {attr_name}");
+'''.format(attr_name=attr_name)
+        else:
+            inner_code = '''
+node.attribute_table.emplace(
+    "{attr_name}", {default_value});
+'''.format(attr_name=attr_name, default_value=default_value)
+
+        attribute_completion_code = '''
+{{
+    auto found = node.attribute_table.find("{attr_name}");
+    if(found == node.attribute_table.end()) {{
+        {code}
+    }}
+}}
+'''.format(attr_name=attr_name, attr_type=attr_type, code=inner_code)
+        attribute_completion_code_list.append(attribute_completion_code)
+
+        attribute_definition = '''
+auto {attr_name} = get<{attr_type}>(node.attribute_table.at("{attr_name}"));
+'''.format(attr_name=attr_name, attr_type=attr_type)
+        attribute_definition_list.append(attribute_definition)
+    # end for
+
+    template = '''
+if(node.op_type == "{op_name}") {{
+    {preprocess}
+    {attribute_completion_code}
+    {postprocess}
+    {{
+        {attribute_definition}
+        {shape_inference_code}
+    }}
+}}
+else
+'''
+    return template.format(
+        op_name=op_name,
+        preprocess=preprocess,
+        attribute_definition="\n".join(attribute_definition_list),
+        shape_inference_code=shape_inference_code,
+        postprocess=postprocess,
+        attribute_completion_code="\n".join(
+            attribute_completion_code_list))
+
+
+def main():
+    template = """
+#ifndef MENOH_ATTRIBUTE_COMPLETION_AND_SHAPE_INFERENCE_HPP
+#define MENOH_ATTRIBUTE_COMPLETION_AND_SHAPE_INFERENCE_HPP
+/*
+ * This file is generated by {script_name}
+ * Do NOT modify this file directly
+ */
+#include <algorithm>
+#include <cassert>
+#include <string>
+#include <unordered_map>
+
+#include <menoh/array.hpp>
+#include <menoh/model_data.hpp>
+
+namespace menoh_impl {{
+    inline auto complete_attribute_and_inference_shape(
```

Comment:
The name `complete_attribute_and_inference_shape` is a bit strange, because "complete attribute" is verb + noun but "inference shape" is noun + noun.

Reply:
Thanks. I'll fix.